common/parallel/TreeCrawler.go

```diff
@@ -101,29 +154,117 @@ func (c *crawler) workerLoop(ctx context.Context, wg *sync.WaitGroup, workerInde
+	// be allowed to proceed. For reasonable sized directories this should ensure that we remain in MaxObjectIndexerSizeInBytes limit
+	// for most practical scenarios. For large/huge directories we might exceed MaxObjectIndexerSizeInBytes by the size of the largest directory.
+	//
+	MaxObjectIndexerSizeInBytes := int64(c.maxObjectIndexerSizeInGB * bytesInGB)
+	lowWaterMark := (MaxObjectIndexerSizeInBytes * 8) / 10
+
+	highWaterMark := MaxObjectIndexerSizeInBytes
+
+	mapSizeInBytes := c.getObjectIndexerMapSize()
+
+	// Nice sunny morning, press that pedal more.
+	if mapSizeInBytes < lowWaterMark {
+		return
+	}
+
+	//
+	// We are nearing danger zone, start slowing down.
+	//
+	// TODO: The sleep duration may be decided based on how fast the
+	//       objectIndexerMap is seen to be growing.
+	//
+	if mapSizeInBytes < highWaterMark && ctx.Err() == nil {
+		time.Sleep(1 * time.Second)
+		return
+	}
+
+	// In danger zone. Don't proceed any further without dropping the objectIndexerMap size below lowWaterMark.
+	for mapSizeInBytes > lowWaterMark && ctx.Err() == nil {
+		time.Sleep(1 * time.Second)
+		mapSizeInBytes = c.getObjectIndexerMapSize()
+	}
+
+	return
+}
+
+func (c *crawler) processOneDirectoryWithAutoPacer(ctx context.Context, workerIndex int) (bool, error) {
 	const maxQueueDirectories = 1000 * 1000
 	const maxQueueDirsForBreadthFirst = 100 * 1000 // figure is somewhat arbitrary.  Want it big, but not huge
 
 	var toExamine Directory
 	stop := false
 
+	// Before picking a new directory to enumerate, call autoPacerWait() to check if we need to
+	// slow down as the objectIndexer might be getting "too full". If Target Traverser is running
+	// slow causing objectIndexer map to get full, we induce wait to let Target Traverser catch up.
+	// autoPacerWait is applicable in case if it's sync and source traverser.
+	if c.isSync && c.isSource {
+		c.autoPacerWait(ctx)
+	}
+
 	// Acquire a directory to work on
 	// Note that we need explicit locking because there are two
 	// mutable things involved in our decision making, not one. (The two being c.unstartedDirs and c.dirInProgressCount)
 	c.cond.L.Lock()
 	{
-		// wait while there's nothing to do, and another thread might be going to add something
-		for len(c.unstartedDirs) == 0 && c.dirInProgressCount > 0 && ctx.Err() == nil {
-			c.cond.Wait() // temporarily relinquish the lock (just on this line only) while we wait for a Signal/Broadcast
+		if c.isSync && !c.isSource {
+			// This is sync and target traverser case. If there is no entry in c.unstartedDirs, wait while source traverser enumerate dirs
+			// and add to tqueue. Once source traverser done with enumeration and close the tqueue, which tells source done with enumeration
+			// and there will be no more dirs.We reset the flag mayHaveMoreDirs in that case.
+			// Othwerwise go ahead process dirs in c.unstartedDirs.
```

Comment:
consider this

	//
	// This is the target traverser of the sync process.
	// It is handled separately since it depends on source traverser to provide it the directories
	// to process. Others do the normal breadth-first enumeration by processing directories they
	// have discovered. 
	// So, for the target traverser case, we have the following wait criteria:
	// 1. unstartedDirs is empty. This is where it will get the next directory to process.
	// 2. mayHaveMoreDirs is true, which means we are not yet done reading from tqueue and
	//    hence more new directories may be added by source traverser.
	//